lib/tapioca/dsl/compilers/active_record_delegated_types.rb

```diff
@@ -0,0 +1,160 @@
+      #     sig { returns(String) }
+      #     def entryable_name; end
+      #
+      #     sig { returns(T::Boolean) }
+      #     def message?; end
+      #
+      #     sig { returns(T.nilable(Message)) }
+      #     def message; end
+      #
+      #     sig { returns(T.nilable(Integer)) }
+      #     def message_id; end
+      #
+      #     sig { returns(T::Boolean) }
+      #     def comment?; end
+      #
+      #     sig { returns(T.nilable(Comment)) }
+      #     def comment; end
+      #
+      #     sig { returns(T.nilable(Integer)) }
+      #     def comment_id; end
+      #   end
+      # end
+      #
+      # ~~~
+      class ActiveRecordDelegatedTypes < Compiler
+        extend T::Sig
+        include Helpers::ActiveRecordConstantsHelper
+
+        ConstantType = type_member { { fixed: T.all(T.class_of(ActiveRecord::Base), Extensions::ActiveRecord) } }
+
+        sig { override.void }
+        def decorate
+          return if constant.__tapioca_delegated_types.nil?
+
+          root.create_path(constant) do |model|
+            model.create_module(DelegatedTypesModuleName) do |mod|
+              constant.__tapioca_delegated_types.each do |role, data|
+                types = data.fetch(:types)
+                options = data.fetch(:options, {})
+                populate_role_accessors(mod, role, types)
+                populate_type_helpers(mod, role, types, options)
+              end
+            end
+
+            model.create_include(DelegatedTypesModuleName)
+          end
+        end
+
+        class << self
+          extend T::Sig
+
+          sig { override.returns(T::Enumerable[Module]) }
+          def gather_constants
+            descendants_of(::ActiveRecord::Base).reject(&:abstract_class?)
+          end
+        end
+
+        private
+
+        sig { params(mod: RBI::Scope, role: Symbol, types: T::Array[String]).void }
+        def populate_role_accessors(mod, role, types)
+          mod.create_method(
+            "#{role}_name",
+            parameters: [],
+            return_type: "String",
+          )
+
+          mod.create_method(
+            "#{role}_class",
+            parameters: [],
+            return_type: "Class",
+          )
+
+          mod.create_method(
+            "build_#{role}",
+            parameters: [create_rest_param("args", type: "T.untyped")],
+            return_type: "T.any(#{types.join(", ")})",
+          )
+        end
+
+        sig { params(mod: RBI::Scope, role: Symbol, types: T::Array[String], options: T::Hash[Symbol, T.untyped]).void }
+        def populate_type_helpers(mod, role, types, options)
+          types.each do |type|
+            populate_type_helper(mod, role, type, options: options)
+          end
+        end
+
+        sig { params(mod: RBI::Scope, role: Symbol, type: String, options: T::Hash[Symbol, T.untyped]).void }
+        def populate_type_helper(mod, role, type, options)
+          singular   = type.tableize.tr("/", "_").singularize
```

Comment:
Don't we have to require `ActiveSupport` at the top to use these methods in a non tapioca development environment?